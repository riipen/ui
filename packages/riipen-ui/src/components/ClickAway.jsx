import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";

class ClickAway extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * Callback fired when the component requests to be closed.
     *
     * @param {object} event The event source of the callback.
     */
    onClickAway: PropTypes.func.isRequired
  };

  static contextType = ThemeContext;

  handleClick = e => {
    e.stopPropagation();
    e.preventDefault();
  };

  handleClickAway = e => {
    const { onClickAway } = this.props;

    onClickAway(e);
  };

  render() {
    const { children } = this.props;

    const theme = this.context;

    const childrenWithProps = React.Children.map(children, child =>
      React.cloneElement(child, { onClick: this.handleClick })
    );

    return (
      <React.Fragment>
        <div onClick={this.handleClickAway}>{childrenWithProps}</div>
        <style jsx>{`
          div {
            bottom: 0;
            left: 0;
            position: fixed;
            right: 0;
            top: 0;
            z-index: ${theme.zIndex.middle};
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default ClickAway;
