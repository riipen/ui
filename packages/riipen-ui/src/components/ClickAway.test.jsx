import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import ClickAway from "./ClickAway";

describe("<ClickAway>", () => {
  it("renders without errors", () => {
    let error;
    const onClickAway = jest.fn();

    try {
      mount(<ClickAway onClickAway={onClickAway} />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const onClickAway = jest.fn();

    const wrapper = mount(<ClickAway onClickAway={onClickAway} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("renders given non null children nodes", () => {
      const child = <span>cheesecakes</span>;
      const onClickAway = jest.fn();

      const wrapper = mount(
        <ClickAway onClickAway={onClickAway}>
          {null}
          {child}
        </ClickAway>
      );

      const childrenNodes = wrapper
        .find("ClickAway")
        .childAt(0)
        .children();
      expect(childrenNodes.at(0).containsMatchingElement(child)).toEqual(true);
    });

    it("sets onClick prop of children node", () => {
      const child = <span>cheesecakes</span>;
      const onClickAway = jest.fn();

      const wrapper = mount(
        <ClickAway onClickAway={onClickAway}>{child}</ClickAway>
      );

      expect(
        wrapper
          .find("ClickAway")
          .find("span")
          .props().onClick
      ).toBeDefined();
    });
  });

  describe("onClickAway prop", () => {
    it("invokes onClickAway when div node is clicked", () => {
      const child = <span>cheesecakes</span>;
      const onClickAway = jest.fn();

      const wrapper = mount(
        <ClickAway onClickAway={onClickAway}>{child}</ClickAway>
      );

      wrapper
        .find("ClickAway")
        .find("div")
        .simulate("click");

      expect(onClickAway).toHaveBeenCalledTimes(1);
    });

    it("does not invoke onClickAway when children node is clicked", () => {
      const child = <span>cheesecakes</span>;
      const onClickAway = jest.fn();

      const wrapper = mount(
        <ClickAway onClickAway={onClickAway}>{child}</ClickAway>
      );

      wrapper
        .find("ClickAway")
        .find("div")
        .childAt(0)
        .simulate("click");

      expect(onClickAway).not.toHaveBeenCalled();
    });
  });
});
