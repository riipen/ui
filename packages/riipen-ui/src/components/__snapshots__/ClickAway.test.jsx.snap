// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClickAway> renders correct snapshot 1`] = `
<withClasses(ClickAway)
  onClickAway={[MockFunction]}
>
  <ClickAway
    classes={
      Array [
        "riipen",
        "riipen-clickaway",
      ]
    }
    onClickAway={[MockFunction]}
  >
    <div
      className="jsx-3334867826"
      onClick={[Function]}
    />
    <JSXStyle
      dynamic={
        Array [
          3,
        ]
      }
      id="1169064260"
    />
  </ClickAway>
</withClasses(ClickAway)>
`;
